packages/jaeger-ui/src/components/TracePage/TraceTimelineViewer/SpanTreeOffset.js

```diff
@@ -14,33 +14,145 @@
+   *     the element the user is now hovering.
+   * @param {string} ancestorId - The span id that the user was hovering over.
+   */
+  handleMouseLeave = (event: SyntheticMouseEvent<HTMLSpanElement>, ancestorId: string) => {
+    if (
+      !(event.relatedTarget instanceof HTMLSpanElement) ||
+      // !event.relatedTarget.getAttribute ||
+      event.relatedTarget.getAttribute('data--ancestor-id') !== ancestorId
+    ) {
+      this.props.removeHoverIndentGuideId(ancestorId);
+    }
+  };
+
+  /**
+   * If the mouse entered this span from anywhere except another span with the same ancestor id, this span's
+   * ancestorId is added to the set of hoverIndentGuideIds.
+   *
+   * @param {Object} event - React Synthetic event tied to mouseenter. Includes the related target which is
+   *     the last element the user was hovering.
+   * @param {string} ancestorId - The span id that the user is now hovering over.
+   */
+  handleMouseEnter = (
+    event: SyntheticMouseEvent<HTMLSpanElement> & { relatedTarget?: { getAttribute: string => string } },
+    ancestorId: string
+  ) => {
+    if (
+      !(event.relatedTarget instanceof HTMLSpanElement) ||
+      // !event.relatedTarget.getAttribute ||
+      event.relatedTarget.getAttribute('data--ancestor-id') !== ancestorId
+    ) {
+      this.props.addHoverIndentGuideId(ancestorId);
+    }
+  };
+
+  render() {
+    const { hasChildren, childrenVisible, onClick } = this.props;
+    const wrapperProps = hasChildren ? { onClick, role: 'switch', 'aria-checked': childrenVisible } : null;
+    const icon = hasChildren && (childrenVisible ? <IoIosArrowDown /> : <IoChevronRight />);
+    return (
+      <span className={`SpanTreeOffset ${hasChildren ? 'is-parent' : ''}`} {...wrapperProps}>
+        {this.ancestorIds.map(ancestorId => (
+          <span
+            key={ancestorId}
+            className={cx('SpanTreeOffset--indentGuide', {
+              'is-active': this.props.hoverIndentGuideIds.has(ancestorId),
+            })}
+            data--ancestor-id={ancestorId}
+            onMouseEnter={event => this.handleMouseEnter(event, ancestorId)}
+            onMouseLeave={event => this.handleMouseLeave(event, ancestorId)}
+          />
+        ))}
+        {icon && (
+          <span
+            className="SpanTreeOffset--iconWrapper"
+            onMouseEnter={event => this.handleMouseEnter(event, this.props.span.spanID)}
+            onMouseLeave={event => this.handleMouseLeave(event, this.props.span.spanID)}
+          >
+            {icon}
+          </span>
+        )}
+      </span>
+    );
+  }
 }
 
-SpanTreeOffset.defaultProps = {
-  hasChildren: false,
-  childrenVisible: false,
-  onClick: null,
-};
+export function mapStateToProps(
+  state: ReduxState,
+  ownProps: SpanTreeOffsetOwnPropsType
+): SpanTreeOffsetPropsType {
+  const hoverIndentGuideIds = state.traceTimeline.hoverIndentGuideIds;
+  return { hoverIndentGuideIds, ...ownProps };
```

Comment:
Use of `ownProps` is optional, and you don't need to include it in the return value:

>You do not need to include values from `ownProps` in the object returned from `mapStateToProps`. `connect` will automatically merge those different prop sources into a final set of props.
>– [react-redux docs](https://react-redux.js.org/using-react-redux/connect-mapstate#ownprops-optional)

Reply:
good point. at Lifion this was an extensive back and forth on clarity vs relying on the API. I prefer relying on the API. removed